File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/drive/advanced/subsystems/LEDStrip.java
package org.firstinspires.ftc.teamcode.drive.advanced.subsystems;

import com.qualcomm.hardware.rev.RevBlinkinLedDriver;
import com.qualcomm.robotcore.hardware.HardwareMap;

import org.firstinspires.ftc.robotcore.internal.system.Deadline;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

public class LEDStrip {
    //    Flashes per second
    public static int flashRate = 2;
    public static Deadline timeout = new Deadline(1000 / flashRate, TimeUnit.MILLISECONDS);
    final RevBlinkinLedDriver driver;
    final List<RevBlinkinLedDriver.BlinkinPattern> colours = new ArrayList<>();
    int index = 0;

    public LEDStrip(HardwareMap hardwareMap, String name) {
        this.driver = hardwareMap.get(RevBlinkinLedDriver.class, name);
    }

    public void update() {
        if (timeout.hasExpired()) {
            timeout.reset();
            if (!colours.isEmpty()) {
                if (index < colours.size() - 1) {
                    driver.setPattern(colours.get(++index));
                } else {
                    driver.setPattern(colours.get(0));
                    index = 0;
                }
            } else {
                driver.setPattern(RevBlinkinLedDriver.BlinkinPattern.BLACK);
            }
        }
    }

    public void changeTo(RevBlinkinLedDriver.BlinkinPattern pattern) {
        colours.clear();
        colours.add(pattern);
    }

    public void changeTo(List<RevBlinkinLedDriver.BlinkinPattern> patterns) {
        colours.clear();
        colours.addAll(patterns);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/galahlib/mechanisms/LEDStrip.kt
package org.firstinspires.ftc.teamcode.galahlib.mechanisms

import com.acmerobotics.dashboard.telemetry.TelemetryPacket
import com.acmerobotics.roadrunner.Action
import com.qualcomm.hardware.rev.RevBlinkinLedDriver
import com.qualcomm.robotcore.hardware.HardwareMap
import org.firstinspires.ftc.robotcore.internal.system.Deadline
import org.firstinspires.ftc.teamcode.galahlib.actions.Loggable
import org.firstinspires.ftc.teamcode.galahlib.actions.LoggableAction
import org.firstinspires.ftc.teamcode.galahlib.actions.doWhile
import java.util.concurrent.TimeUnit

class LEDStrip(hardwareMap: HardwareMap, name: String) {
    val driver: RevBlinkinLedDriver = hardwareMap.get(RevBlinkinLedDriver::class.java, name)

    fun showPattern(
        colors: List<RevBlinkinLedDriver.BlinkinPattern>,
        blinksPerSecond: Double
    ): Action {
        return object : Action {
            val timeout = Deadline((1000 / blinksPerSecond).toLong(), TimeUnit.MILLISECONDS)
            var index = 0
            override fun run(p: TelemetryPacket): Boolean {
                if (timeout.hasExpired()) {
                    if (index++ >= colors.size) index = 0

                    driver.setPattern(colors[index])
                }

                return true
            }
        }
    }

    fun showWhile(
        colors: List<RevBlinkinLedDriver.BlinkinPattern>,
        blinksPerSecond: Double,
        action: LoggableAction
    ): LoggableAction {
        return doWhile(
            Loggable("SHOW_PATTERN", showPattern(colors, blinksPerSecond)),
            action
        )
    }
}
